core/primitives/src/shard_layout.rs

```diff
@@ -325,6 +325,20 @@ pub struct ShardUId {
     pub shard_id: u32,
 }
 
+/// A wrapper for ShardUId used to denote when we expect the mapped version of the value.
+///
+/// ReshardingV3 uses mapping strategy for the State column.
+/// It means we access trie storage using a mapped shard UID, that can be the same shard UID,
+/// resharding parent shard UID, or any ancestor shard UID in the tree of resharding history.
+#[derive(Clone, Copy)]
+pub struct MappedShardUId(pub ShardUId);
```

Comment:
Nice to have this wrapper to clearly distinguish between mapped and non-mapped ShardUId :)